File: mediapipe/tasks/web/components/processors/detection_result.ts
/**
 * Copyright 2023 The MediaPipe Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Detection as DetectionProto} from '../../../../framework/formats/detection_pb';
import {Detection} from '../../../../tasks/web/components/containers/detection_result';

const DEFAULT_CATEGORY_INDEX = -1;

/** Converts a Detection proto into a Detection object. */
export function convertFromDetectionProto(source: DetectionProto): Detection {
  const scores = source.getScoreList();
  const indexes = source.getLabelIdList();
  const labels = source.getLabelList();
  const displayNames = source.getDisplayNameList();

  const detection: Detection = {categories: [], keypoints: []};
  for (let i = 0; i < scores.length; i++) {
    detection.categories.push({
      score: scores[i],
      index: indexes[i] ?? DEFAULT_CATEGORY_INDEX,
      categoryName: labels[i] ?? '',
      displayName: displayNames[i] ?? '',
    });
  }

  const boundingBox = source.getLocationData()?.getBoundingBox();
  if (boundingBox) {
    detection.boundingBox = {
      originX: boundingBox.getXmin() ?? 0,
      originY: boundingBox.getYmin() ?? 0,
      width: boundingBox.getWidth() ?? 0,
      height: boundingBox.getHeight() ?? 0,
      angle: 0.0,
    };
  }

  if (source.getLocationData()?.getRelativeKeypointsList().length) {
    for (const keypoint of
             source.getLocationData()!.getRelativeKeypointsList()) {
      detection.keypoints.push({
        x: keypoint.getX() ?? 0.0,
        y: keypoint.getY() ?? 0.0,
        score: keypoint.getScore() ?? 0.0,
        label: keypoint.getKeypointLabel() ?? '',
      });
    }
  }

  return detection;
}

File: mediapipe/tasks/cc/components/containers/detection_result.cc
/* Copyright 2022 The MediaPipe Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "mediapipe/tasks/cc/components/containers/detection_result.h"

#include <optional>
#include <string>
#include <vector>

#include "mediapipe/framework/formats/detection.pb.h"
#include "mediapipe/framework/formats/location_data.pb.h"
#include "mediapipe/tasks/cc/components/containers/category.h"
#include "mediapipe/tasks/cc/components/containers/rect.h"

namespace mediapipe::tasks::components::containers {

constexpr int kDefaultCategoryIndex = -1;

Detection ConvertToDetection(const mediapipe::Detection& detection_proto) {
  Detection detection;
  for (int idx = 0; idx < detection_proto.score_size(); ++idx) {
    detection.categories.push_back(
        {/* index= */ detection_proto.label_id_size() > idx
             ? detection_proto.label_id(idx)
             : kDefaultCategoryIndex,
         /* score= */ detection_proto.score(idx),
         /* category_name */ detection_proto.label_size() > idx
             ? detection_proto.label(idx)
             : "",
         /* display_name */ detection_proto.display_name_size() > idx
             ? detection_proto.display_name(idx)
             : ""});
  }
  Rect bounding_box;
  if (detection_proto.location_data().has_bounding_box()) {
    mediapipe::LocationData::BoundingBox bounding_box_proto =
        detection_proto.location_data().bounding_box();
    bounding_box.left = bounding_box_proto.xmin();
    bounding_box.top = bounding_box_proto.ymin();
    bounding_box.right = bounding_box_proto.xmin() + bounding_box_proto.width();
    bounding_box.bottom =
        bounding_box_proto.ymin() + bounding_box_proto.height();
  }
  detection.bounding_box = bounding_box;
  if (!detection_proto.location_data().relative_keypoints().empty()) {
    detection.keypoints = std::vector<NormalizedKeypoint>();
    detection.keypoints->reserve(
        detection_proto.location_data().relative_keypoints_size());
    for (const auto& keypoint :
         detection_proto.location_data().relative_keypoints()) {
      detection.keypoints->push_back(
          {keypoint.x(), keypoint.y(),
           keypoint.has_keypoint_label()
               ? std::make_optional(keypoint.keypoint_label())
               : std::nullopt,
           keypoint.has_score() ? std::make_optional(keypoint.score())
                                : std::nullopt});
    }
  }
  return detection;
}

DetectionResult ConvertToDetectionResult(
    std::vector<mediapipe::Detection> detections_proto) {
  DetectionResult detection_result;
  detection_result.detections.reserve(detections_proto.size());
  for (const auto& detection_proto : detections_proto) {
    detection_result.detections.push_back(ConvertToDetection(detection_proto));
  }
  return detection_result;
}
}  // namespace mediapipe::tasks::components::containers
